spec/truffle/arguments_descriptor_spec.rb

```diff
@@ -0,0 +1,197 @@
+# truffleruby_primitives: true
+
+require_relative '../ruby/spec_helper'
+
+module ArgumentsDescriptorSpecs
+  def self.tr?
+    RUBY_ENGINE == 'truffleruby'
+  end
+
+  unless tr?
+    module Primitive
+      def self.arguments_descriptor
+        []
+      end
+
+      def self.arguments
+        []
+      end
+    end
+  end
+
+  def self.no_kws(a, b)
+    [a, b, Primitive.arguments_descriptor, Primitive.arguments]
+  end
+
+  def self.only_kws(a:, b:)
+    [a, b, Primitive.arguments_descriptor, Primitive.arguments]
+  end
+
+  def self.only_kws_and_opt(a:, b: 101)
+    [a, b, Primitive.arguments_descriptor, Primitive.arguments]
+  end
+
+  def self.posn_kws(a, b:)
+    [a, b, Primitive.arguments_descriptor, Primitive.arguments]
+  end
+
+  def self.posn_kws_defaults(a, b: 101, c: 102)
+    [a, b, c, Primitive.arguments_descriptor, Primitive.arguments]
+  end
+
+  def self.opt_and_kws(a, b=2, c: nil)
+    [a, b, c, Primitive.arguments_descriptor, Primitive.arguments]
+  end
+
+  def self.struct_new_like(a, *b, c: 101)
+    [a, b, c, Primitive.arguments_descriptor, Primitive.arguments]
+  end
+
+  def self.describe_like(a, b = nil)
+    [a, b, Primitive.arguments_descriptor, Primitive.arguments]
+  end
+
+  def self.single(a)
+    [a, Primitive.arguments_descriptor, Primitive.arguments]
+  end
+
+  def self.rest(*a)
+    [a, Primitive.arguments_descriptor, Primitive.arguments]
+  end
+
+  def self.kw_rest(**a)
+    [a, Primitive.arguments_descriptor, Primitive.arguments]
+  end
+
+  def self.kw_and_kw_rest(a:, **b)
+    [a, b, Primitive.arguments_descriptor, Primitive.arguments]
+  end
+
+  def self.mixture(a, b = nil, c = nil, d, e: nil, **f)
+    [a, b, c, d, e, f, Primitive.arguments_descriptor, Primitive.arguments]
+  end
+
+  class NewKW
+    def self.new(a:, b:)
+      [a, b, Primitive.arguments_descriptor, Primitive.arguments]
+    end
+  end
+
+  class NewRest
+    def self.new(*a)
+      [a, Primitive.arguments_descriptor, Primitive.arguments]
+    end
+  end
+end
+
+describe "Arguments descriptors" do
+  it "are empty for simple calls" do
+    ArgumentsDescriptorSpecs.no_kws(1, 2).should == [1, 2, [], ArgumentsDescriptorSpecs.tr? ? [1, 2]: []]
+  end
+
+  it "contain keyword arguments" do
+    ArgumentsDescriptorSpecs.only_kws(a: 1, b: 2).should == [1, 2, ArgumentsDescriptorSpecs.tr? ? [:a, :b, 0, true] : [], ArgumentsDescriptorSpecs.tr? ? [{a: 1, b: 2}] : []]
```

Comment:
This is not readable, how about always ignoring the last 2 values on CRuby?
Could be like:
```ruby
engine = TruffleRuby ? (0..-1) : (0..-3)
ArgumentsDescriptorSpecs.only_kws(a: 1, b: 2)[engine].should == [1, 2, [:a, :b, 0, true], [{a: 1, b: 2}]][engine]
```

Alternatively, since this is a spec/truffle there is no need to make it work on CRuby, so maybe just don't try to run this on CRuby?

Reply:
Could also be:
```ruby
check ArgumentsDescriptorSpecs.only_kws(a: 1, b: 2), [1, 2, [:a, :b, 0, true], [{a: 1, b: 2}]]
```
and define `check` in the `describe`

Reply:
> so maybe just don't try to run this on CRuby?

It also checks the actual handling of the arguments (which is why two are tagged on TruffleRuby.) But I could split them up.

Reply:
It's probably useful to let it run on CRuby to check expectations are correct and match CRuby behavior, but it needs less noise.

Reply:
Split it up to test one part per line.